2018/001-external-authentication.md

```diff
@@ -0,0 +1,109 @@
+
+Dex supports many authentication "connectors" - LDAP, SAML 2.0, OpenID Connect,
+as well as several less common connectors such as GitHub, LinkedIn, etc. Dex
+additionally supports configuring multiple connectors at once, allowing users
+to choose which service they need to authenticate against.
+
+This RFC proposes allowing Kubic operators to configure arbitary additional
+connectors within Kubic, pointing towards their own authentication backend.
+Initially, this will be limited to Dex's LDAP, SAML 2.0 and OpenID Connect
+connectors initially.
+
+This RFC does not propose removing that managed LDAP server, though that may
+be something we wish to persue in the future.
+
+## Detailed RFC
+
+### Velum Workflow Changes
+
+We propose to leave the initial bootstrap workflow unchanged, and instead
+allow additional authentication connectors to be configured post-bootstap, in
+a similar manor to how we allow containter registries to be configured
+post-bootstap.
+
+Operators should be able to login to Velum, navigate to the Settings tab, and
+view, add or remove authentication connectors. Each authentication connector
+type (LDAP, SAML 2.0, OpenID Connect) will have a different view for the view
+and add pages, as Dex requires a different set of configuration inputs for
+each.
+
+It should be possible for operators to confgiure multiple connectors
+simultaneously, including multiple connectors of the same type (e.g. 2x LDAP)
+simultaneously.
+
+Once connectors have been configured, Velum will show an "Apply Changes" button,
+again - similar to containter registries page, and Salt will be called to push
+out the updated configuration.
+
+### Salt Implementation
+
+Velum will store the connector details in it's database, and will render the
+values into the Pillar API endpoint for Salt to read.
+
+Once in salt, most of the pieces are already in place for this. We'll need to
+update the Dex ConfigMap template (https://github.com/kubic-project/salt/blob/292b025681915a6ad4d7378624385d231f2f21d4/salt/addons/dex/manifests/15-configmap.yaml#L21)
+to render the new connector details, and þen salt will apply the change to
+Kubernetes, causing a roll out of the change.
+
+### CaaSP CLI Implementation
+
+As we now support SAML 2.0 and OpenID Connect, the current `caasp-cli` workflow
+will no longer be possible. CaaSP CLI must be updated to fully implement the
+OIDC browser based workflow. When a user logs in with `caasp-cli`, they must be
+presented with a URL to open in their browser, where they will login and copy
+the generated token back to a waiting `caasp-cli` in order to complete the
+authentication flow. The existing expectations that `caasp-cli` can provide the
+username and password on behalf of the user to Dex are no longer valid.
+
+## Concerns and Unresolved Questions
+
+1. Dex will be restarted, in a rolling fashion, as the configmap is updated.
+   This may introduce authentication issues during the change.
```

Comment:
Given dex is running as a deployment I think kubernetes should be capable of handling the update without introducing downtime... isn't it?

Reply:
In theory, yes... In reality, we need to take care to ensure this is working as expected, and the correct healthchecks etc are in place to make it go smoothly.